### PR TITLE
Add snapshot performance integration test.

### LINKED ISSUE #2028: Add snapshot create/restore latency integration tests.

### PR DESCRIPTION
**Pull request by sandreim**
## Reason for This PR

Fixes #2028

## Description of Changes

-  Add metrics setup in build_from_snapshot()
-  Add test to sample 3 load and 3 create latencies
-  Add some temporary baseline latencies

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+        # Create a snapshot builder from a microvm.
+        snapshot_builder = SnapshotBuilder(vm)
+        snapshot_builder.create([rw_disk],
+                                ssh_key,
+                                snapshot_type)
+        metrics = vm.flush_metrics(metrics_fifo)
+
+        if snapshot_type == SnapshotType.FULL:
+            value = metrics['latencies_us']['full_create_snapshot'] / 1000
```
Comment (gc-plp) [inline tests/integration_tests/performance/test_snapshot_perf.py:87]: `1000` should be a constant, such as:
`USEC_IN_MSEC = 1000`

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    ssh_key = context.disk.ssh_key()
+    # Create a fresh microvm from aftifacts.
+    basevm = vm_builder.build(kernel=context.kernel,
+                              disks=[rw_disk],
+                              ssh_key=ssh_key,
+                              config=context.microvm,
+                              enable_diff_snapshots=enable_diff_snapshots)
+
+    network_config = net_tools.UniqueIPv4Generator.instance()
+    _, host_ip, guest_ip = basevm.ssh_network_config(network_config,
```
Comment (gc-plp) [inline tests/integration_tests/performance/test_snapshot_perf.py:123]: Could we use pass the parameters by name? Passing `1` is a bit confusing for someone who doesn't know the parameter order.

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                              context.disk.name()))
+
+    # Create a rw copy artifact.
+    rw_disk = context.disk.copy()
+    # Get ssh key from read-only artifact.
+    ssh_key = context.disk.ssh_key()
+
+    # Measure a burst of snapshot create calls.
+    for i in range(SAMPLE_COUNT):
+        # Create a fresh microvm from aftifacts.
```
Comment (gc-plp) [inline tests/integration_tests/performance/test_snapshot_perf.py:61]:
```suggestion
        # Create a fresh microvm from artifacts.
```

**Comment by gc-plp** [comment]
It seems like `test_snapshot_resume_latency` is causing intermittent failures.

We should look into this before merging.

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 2 above)
Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:123]: This is the old framework. I would not invest in fixing the tech debt in this PR. I think we still need to rewrite this in the new framework.

Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:130]: done

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 1 above)
Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py]: done

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 3 above)
Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:63]: done

File: tests/framework/builder.py [hunk 4]
```diff
@@ -59,19 +61,32 @@ def build(self,
...
+                            guest_ip=None,
+                            netmask_len=None,
                             resume=False,
                             # Enable incremental snapshot capability.
                             enable_diff_snapshots=False):
         """Build a microvm from a snapshot artifact."""
         self.init_root_path()
         vm = init_microvm(self.root_path, self.bin_cloner_path)
+        vm.jailer.daemonize = False
+        vm.jailer.extra_args.update({'seccomp-level': "0"})
```
Comment (iulianbarbu) [inline tests/framework/builder.py:80]: Why is this necessary?

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+import host_tools.network as net_tools  # pylint: disable=import-error
+import host_tools.logging as log_tools
+
+
+# How many latencies do we sample per test.
+SAMPLE_COUNT = 3
+USEC_IN_MSEC = 1000
+
+# Latencies in milliseconds.
+CREATE_LATENCIES = {
```
Comment (iulianbarbu) [inline tests/integration_tests/performance/test_snapshot_perf.py:23]: nit:
```suggestion
CREATE_LATENCY_BASELINES = {
```

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        'FULL':  1000,
+        'DIFF':  0,
+    }
+}
+
+# The latencies are pretty high during integration tests and
+# this is tracked here:
+# https://github.com/firecracker-microvm/firecracker/issues/2027
+# TODO: Update the table after fix. Target is < 5ms.
+LOAD_LATENCIES = {
```
Comment (iulianbarbu) [inline tests/integration_tests/performance/test_snapshot_perf.py:38]: nit:
```suggestion
LOAD_LATENCY_BASELINES = {
```

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        if snapshot_type == SnapshotType.FULL:
+            value = metrics['latencies_us']['full_create_snapshot']
+            baseline = CREATE_LATENCIES[context.microvm.name()]['FULL']
+        else:
+            value = metrics['latencies_us']['diff_create_snapshot']
+            baseline = CREATE_LATENCIES[context.microvm.name()]['DIFF']
+
+        value = value / USEC_IN_MSEC
+
+        assert baseline > value, "CreateSnapshot latency degraded."
```
Comment (iulianbarbu) [inline tests/integration_tests/performance/test_snapshot_perf.py:97]: nit:
```suggestion
        assert baseline > value, "CreateSnapshot latency degraded with: {} milliseconds.".format(value - baseline)
```

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            metrics = json.loads(data_point)
+            cur_value = metrics['latencies_us']['load_snapshot'] / USEC_IN_MSEC
+            if cur_value > 0:
+                value = cur_value
+                break
+
+        baseline = LOAD_LATENCIES[context.microvm.name()]
+        logger.info("Latency {}/{}: {} ms".format(i + 1, SAMPLE_COUNT, value))
+        value = value / USEC_IN_MSEC
+        assert baseline > value, "LoadSnapshot latency degraded."
```
Comment (iulianbarbu) [inline tests/integration_tests/performance/test_snapshot_perf.py:183]: nit: add degradation numbers to assert.

Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:183]: It will show up in console.

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 7 above)
Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:97]: assert already prints that information.

Comment (iulianbarbu) [reply tests/integration_tests/performance/test_snapshot_perf.py:97]: Oh, that's right. Cool.

File: tests/framework/builder.py (hunk 4 above)
Comment (sandreim) [reply tests/framework/builder.py:80]: No, just a leftover.

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 5 above)
Comment (sandreim) [reply tests/integration_tests/performance/test_snapshot_perf.py:23]: done

File: tests/framework/builder.py (hunk 4 above)
Comment (sandreim) [reply tests/framework/builder.py:80]: done

**Comment by sandreim** [comment]
@gc-plp @iulianbarbu updated. PTAL!
